core/vm/evm.go

```diff
@@ -186,9 +171,19 @@ type EVM struct {
 	callGasTemp uint64
 }
 
+type ChainConfig interface {
+	IsEIP158(blockNum *big.Int) bool
+	IsSubnetEVM(timestamp uint64) bool
+	IsCancun(blockNum *big.Int, timestamp uint64) bool
+	IsPrecompileEnabled(addr common.Address, timestamp uint64) bool
+	Rules(blockNum *big.Int, timestamp uint64) params.Rules
+}
+
+var NewEVM = newEVM
```

Comment:
This may be a little jank, but we can consider it a shortcut for this pattern:
Convert functions to methods on the struct and reinstate the top-level function Foo() as an alias to DefaultConfig().Foo() ([slog.Info() example](https://cs.opensource.google/go/go/+/refs/tags/go1.23.0:src/log/slog/logger.go;l=291)).
as suggested by @ARR4N

Reply:
I think it makes sense. IMHO globals are really annoying.

Reply:
I'm not sure I follow. The global is introduced so that `core` can substitute it with a different implementation. The `slog` pattern would be more like:

```go
type EVMConfig struct {
//...
}

func (c *EVMConfig) New(...) *EVM {
// ... contents of original constructor, modified to use the `EVMConfig`
}

func DefaultEVMConfig() *EVMConfig {
//...
}

func NewEVM(...) *EVM {
// what used to be here is now in `EVMConfig.New()`
    return DefaultEVMConfig().New()
}
```

Reply:
I switched to this pattern. We still somehow need a factory since the arguments provided are used to build the config object.

Reply:
Requiring a factory is ok; what I'm saying is that it shouldn't be a global variable to allow it to be arbitrarily substituted. The `slog.Info()` pattern unfortunately has a red herring in that you can swap out the default logger; [this is a better example of what I intended](https://github.com/darvaza-proxy/darvaza/blob/e8cb3af76ff47ab319166c60a767c22b7f8b961b/shared/storage/simple/config.go#L17).